### PR TITLE
feat: Add login and generating tokens functionality

### PR DESCRIPTION
**Pull request by abdul-ahad12**

### CODE REVIEW

File: apps/backend/api/src/shared/interface.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ export interface ICustomRequest extends Request {
 export interface ApiResponse<T> {
   success: boolean;
   message: string;
-  data: T | null;
+  data?: T | null;
```
Comment (theshdb) [inline apps/backend/api/src/shared/interface.ts:8]: This is a mandatory field, revert this change. 

If there is no data to send then return `data: null`

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 2]
```diff
@@ -5,35 +5,124 @@ import {
...
-import { Response } from 'express';
+import { Response, Request } from 'express';
+import { UserRoleEnum } from 'src/shared/enums';
+import { logger } from 'src/logger';
+import { ApiResponse } from 'src/shared/interface';
 
 @Controller('auth')
 export class AuthController {
-  constructor(private readonly authService: AuthService) {}
+  constructor(private readonly authService: AuthService) { }
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:23]:
```suggestion
  constructor(private readonly authService: AuthService) {}
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 3]
```diff
@@ -5,35 +5,124 @@ import {
...
 
   @Post('otp/request')
-  @HttpCode(HttpStatus.OK)
-  async requestOtp(
-    @Body() requestOtpDto,
-  ): Promise<{ success: boolean; message: string }> {
+  async requestOtp(@Body() requestOtpDto: { phoneNumber: string }): Promise<ApiResponse<null>> {
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:26]:
```suggestion
  async requestOtp(@Body() number: string ): Promise<ApiResponse<null>> {
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 4]
```diff
@@ -5,35 +5,124 @@ import {
...
     const { phoneNumber } = requestOtpDto;
-    return this.authService.requestOtp(phoneNumber);
+    logger.debug(`Received OTP request for phoneNumber: ${phoneNumber}`);
+    try {
+      await this.authService.requestOtp(phoneNumber);
+      logger.debug('OTP sent successfully');
+
+      return { success: true, message: 'OTP sent successfully.' };
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:33]:
```suggestion
      return { success: true, message: 'OTP sent successfully.', data: null };
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 5]
```diff
@@ -5,35 +5,124 @@ import {
...
-  async requestOtp(
-    @Body() requestOtpDto,
-  ): Promise<{ success: boolean; message: string }> {
+  async requestOtp(@Body() requestOtpDto: { phoneNumber: string }): Promise<ApiResponse<null>> {
     const { phoneNumber } = requestOtpDto;
-    return this.authService.requestOtp(phoneNumber);
+    logger.debug(`Received OTP request for phoneNumber: ${phoneNumber}`);
+    try {
+      await this.authService.requestOtp(phoneNumber);
+      logger.debug('OTP sent successfully');
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:31]:
```suggestion
      logger.debug('AuthController. requestOtp: OTP sent successfully');
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 6]
```diff
@@ -5,35 +5,124 @@ import {
...
+
+      return { success: true, message: 'OTP sent successfully.' };
+    } catch (error) {
+      logger.error(`Error sending OTP: ${error.message}`);
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:35]:
```suggestion
      logger.error(`AuthController. requestOtp: Error sending OTP: ${error.message}`);
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 7]
```diff
@@ -5,35 +5,124 @@ import {
...
+  ): Promise<ApiResponse<null>> {
     const { phoneNumber, otp } = verifyOtpDto;
-    const { success, message, onboardingToken } =
-      await this.authService.signupInitiate(phoneNumber, otp);
-    res.cookie('onboarding_token', onboardingToken, {
-      httpOnly: true,
-      secure: false,
-      maxAge: 60 * 24 * 60 * 60 * 1000,
-    });
+    logger.debug(`Verifying OTP for phoneNumber: ${phoneNumber}`);
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:48]:
```suggestion
    logger.debug(`AuthController. verifyOtp: Verifying OTP for phoneNumber: ${phoneNumber}`);
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 8]
```diff
@@ -5,35 +5,124 @@ import {
...
+    try {
+      const response = await this.authService.signupInitiate(phoneNumber, otp);
+      if (response.success) {
+        logger.debug('OTP verification successful, setting onboarding token cookie');
+        res.setHeader('x-onboarding-token', response.data.onboardingToken);
+      }
+      return { success: response.success, message: response.message };
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:55]:
```suggestion
      return { success: response.success, message: response.message, data:null };
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 9]
```diff
@@ -5,35 +5,124 @@ import {
...
-    res.cookie('onboarding_token', onboardingToken, {
-      httpOnly: true,
-      secure: false,
-      maxAge: 60 * 24 * 60 * 60 * 1000,
-    });
+    logger.debug(`Verifying OTP for phoneNumber: ${phoneNumber}`);
+    try {
+      const response = await this.authService.signupInitiate(phoneNumber, otp);
+      if (response.success) {
+        logger.debug('OTP verification successful, setting onboarding token cookie');
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:52]:
```suggestion
        logger.debug('AuthController.verifyOtp: OTP verification successful, setting onboarding token cookie');
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 10]
```diff
@@ -5,35 +5,124 @@ import {
...
+        res.setHeader('x-onboarding-token', response.data.onboardingToken);
+      }
+      return { success: response.success, message: response.message };
+    } catch (error) {
+      logger.error(`Error verifying OTP: ${error.message}`);
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:57]:
```suggestion
      logger.error(`AuthController.verifyOtp: Error verifying OTP: ${error.message}`);
```

File: apps/backend/api/src/modules/auth/auth.controller.ts [hunk 11]
```diff
@@ -5,35 +5,124 @@ import {
...
+        'Failed to send OTP. Please try again later.'
+      );
+    }
   }
 
   @Post('otp/verify')
-  @HttpCode(HttpStatus.OK)
-  async verifyOtp(@Body() verifyOtpDto, @Res() res: Response): Promise<void> {
+  async verifyOtp(
+    @Body() verifyOtpDto: { phoneNumber: string; otp: string },
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.controller.ts:44]:
```suggestion
    @Body() body: { phoneNumber: string; otp: string },
```

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 12]
```diff
@@ -17,69 +28,95 @@ export class AuthService {
...
-      success: true,
-      message: 'Signup initiated successfully',
-      onboardingToken,
-    };
+    const onboardingToken = await this.tokenService.generateOnboardingToken(phoneNumber);
+
+    return { success: true, message: 'Signup initiated successfully', data: { onboardingToken } };
+  }  
+
+  async login(phoneNumber: string, otp: string, role: UserRoleEnum): Promise<any> {
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:68]:
```suggestion
  async login(phoneNumber: string, otp: string, role: UserRoleEnum): Promise<dine the custom type or the return object here> {
```

**Review by shettydev — APPROVED** [review]

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 13]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
-      success: true,
-      message: 'Signup initiated successfully',
-      onboardingToken,
-    };
+    const onboardingToken = await this.tokenService.generateOnboardingToken(phoneNumber);
+
+    return { success: true, message: 'Signup initiated successfully', data: { onboardingToken } };
+  }
+
+  async login(phoneNumber: string, otp: string, role: UserRoleEnum): Promise<ApiResponse<{ accessToken: string; refreshToken: string }>> {
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:66]:
```suggestion
  async login(phoneNumber: string, otp: string, role: UserRoleEnum): Promise<{ accessToken: string; refreshToken: string }>{
```

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 14]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
+    return { success: true, message: 'Login successful', data: { accessToken, refreshToken } };
   }
 
-  decryptOnboardingToken(token: string): string {
-    const payload = this.jwtService.verify(token, {
-      secret: this.configService.get<string>('ONBOARDING_JWT_SECRET'),
-    });
-    return this.cryptoService.decrypt(payload.phone);
+  async refreshToken(refreshToken: string): Promise<ApiResponse<{ accessToken: string; refreshToken: string }>> {
+    logger.debug('AuthService. refreshToken: Refreshing token');
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:95]:
```suggestion
```

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 15]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
+    try {
+      const payload = this.jwtService.verify(refreshToken, {
+        secret: this.configService.get<string>('JWT_REFRESH_SECRET'),
+      });
+
+      const user = await dbRepo(User).findOne({ where: { id: payload.userId } });
+
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:102]:
```suggestion
```

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 16]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
+      throw new InternalServerErrorException('Twilio credentials are not set');
     }
-
     this.twilioClient = new Twilio(accountSid, authToken);
   }
 
-  async requestOtp(
-    phoneNumber: string,
-  ): Promise<{ success: boolean; message: string }> {
+  async requestOtp(phoneNumber: string): Promise<ApiResponse<null>> {
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:42]:
```suggestion
  async requestOtp(phoneNumber: string): Promise<void> {
```

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 17]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
-    phoneNumber: string,
-    otp: string,
-  ): Promise<{
-    success: boolean;
-    message: string;
-    onboardingToken: string;
-  }> {
-    const existingUser =
-      this.configService.get<string>('ENVIRONMENT') === 'local' && false; // Temporarily set to false for testing as user.service is not defined
+  async signupInitiate(phoneNumber: string, otp: string): Promise<ApiResponse<{ onboardingToken: string }>> {
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:51]:
```suggestion
  async signupInitiate(phoneNumber: string, otp: string): Promise<string> {
```

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 18]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
-    const payload = this.jwtService.verify(token, {
-      secret: this.configService.get<string>('ONBOARDING_JWT_SECRET'),
-    });
-    return this.cryptoService.decrypt(payload.phone);
+  async refreshToken(refreshToken: string): Promise<ApiResponse<{ accessToken: string; refreshToken: string }>> {
+    logger.debug('AuthService. refreshToken: Refreshing token');
+    try {
+      const payload = this.jwtService.verify(refreshToken, {
+        secret: this.configService.get<string>('JWT_REFRESH_SECRET'),
+      });
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:99]: What if the verification fails?

File: apps/backend/api/src/modules/auth/auth.service.ts [hunk 19]
```diff
@@ -17,69 +27,96 @@ export class AuthService {
...
+
+    const accessToken = await this.tokenService.generateAccessToken(user.id, user.phoneNumber, user.role);
+    const refreshToken = await this.tokenService.generateRefreshToken(user.id);
+
+    await this.redisService.set(
+      `refreshToken:${user.id}`,
+      refreshToken,
+      'EX',
+      60 * 60 * 24 * 7,
+    );
```
Comment (theshdb) [inline apps/backend/api/src/modules/auth/auth.service.ts:89]: where are we using this stored value in redis?
